### PR TITLE
Documentation for WFST decoding

### PR DESCRIPTION
**Pull request by anand-nv**
Documentation for Class LM based WFST decoding and Deployment of WFST decoder

### CODE REVIEW

**Comment by rmittal-github** [comment]
@anand-nv please move this to release/2.12.0 branch.

**Comment by anand-nv** [comment]
> @anand-nv please move this to release/2.12.0 branch.

Done

**Comment by rmittal-github** [comment]
please review @LynseyFabel, this is needed in the release.
@anand-nv there should be a link to these tutorials in the Riva docs on the left pane under `Tutorials -> Speech Recognition` section, right?

File: wfst_class_lm.md [hunk 1]
```diff
@@ -0,0 +1,159 @@
...
+This can be achieved in the following two ways:
+    - Build the language model from text containing standard text as well as
+  word class labels
+    - n-gram LM interpolation/mixing of two LMs, one from standard text and other 
+  from domain specific grammars.
+
+The latter option provides for better weighting between in-domain and out of 
+domain LMs. 
+Please refer to the tutorial ["How To Train, Evaluate, and Fine-Tune an n-gram Language Model"](https://github.com/nvidia-riva/tutorials/blob/main/asr-python-advanced-nemo-ngram-training-and-finetuning.ipynb) 
+for LM model interpolation using kenlm.
```
Comment (LynseyFabel) [inline wfst_class_lm.md:70]:
```suggestion
for LM model interpolation using KenLM. 
```

File: wfst_class_lm.md [hunk 2]
```diff
@@ -0,0 +1,159 @@
...
+    - Build the language model from text containing standard text as well as
+  word class labels
+    - n-gram LM interpolation/mixing of two LMs, one from standard text and other 
+  from domain specific grammars.
+
+The latter option provides for better weighting between in-domain and out of 
+domain LMs. 
+Please refer to the tutorial ["How To Train, Evaluate, and Fine-Tune an n-gram Language Model"](https://github.com/nvidia-riva/tutorials/blob/main/asr-python-advanced-nemo-ngram-training-and-finetuning.ipynb) 
+for LM model interpolation using kenlm. 
+Alternatively [SRILM](http://www.speech.sri.com/projects/srilm/) (license required) can also be used here
```
Comment (LynseyFabel) [inline wfst_class_lm.md:71]:
```suggestion
Alternatively, [SRILM](http://www.speech.sri.com/projects/srilm/) (license required) can also be used here.
```

File: wfst_class_lm.md [hunk 3]
```diff
@@ -0,0 +1,159 @@
...
+
+The process for supporting word classes in the WFST framework basically involves
+replacement of the arcs representing the class labels, with FSTs created from 
+the entities present in the word classes. Given a `G.fst` containing 
+word class labels (`#entity:<class>`), and the files (filenames `<class>.txt`) 
+containing the list of entities for the class. The list of entities  is not
```
Comment (LynseyFabel) [inline wfst_class_lm.md:77]:
```suggestion
containing the list of entities for the class. The list of entities is not 
```

File: wfst_class_lm.md [hunk 4]
```diff
@@ -0,0 +1,159 @@
...
+Please refer to the tutorial ["How To Train, Evaluate, and Fine-Tune an n-gram Language Model"](https://github.com/nvidia-riva/tutorials/blob/main/asr-python-advanced-nemo-ngram-training-and-finetuning.ipynb) 
+for LM model interpolation using kenlm. 
+Alternatively [SRILM](http://www.speech.sri.com/projects/srilm/) (license required) can also be used here
+
+The process for supporting word classes in the WFST framework basically involves
+replacement of the arcs representing the class labels, with FSTs created from 
+the entities present in the word classes. Given a `G.fst` containing 
+word class labels (`#entity:<class>`), and the files (filenames `<class>.txt`) 
+containing the list of entities for the class. The list of entities  is not 
+restricted to words and could be phrases
```
Comment (LynseyFabel) [inline wfst_class_lm.md:78]:
```suggestion
restricted to words and could be phrases.
```

File: wfst_class_lm.md [hunk 5]
```diff
@@ -0,0 +1,159 @@
...
+
+For example, in the above example `cities.txt` would contain at the very least the
```
Comment (LynseyFabel) [inline wfst_class_lm.md:80]:
```suggestion
For example, in the above example, `cities.txt` would contain at the very least the 
```

File: wfst_class_lm.md [hunk 6]
```diff
@@ -0,0 +1,159 @@
...
+following
+```shell
+$> cat city.txt
+london
+new york
+paris
+seattle
+```
+
+## Process to create the `TLG.fst` supporting word classes
```
Comment (LynseyFabel) [inline wfst_class_lm.md:90]:
```suggestion
## Process to Create `TLG.fst` Supporting Word Classes
```

File: wfst_class_lm.md [hunk 7]
```diff
@@ -0,0 +1,159 @@
...
+Checkout the `class_lm` branch from https://github.com/anand-nv/riva-asrlib-decoder/tree/class_lm. 
+All the below scripts are available in `src/riva/decoder/scripts/prepare_TLG_fst` path.
```
Comment (LynseyFabel) [inline wfst_class_lm.md:92]:
```suggestion
The following scripts are available in the `src/riva/decoder/scripts/prepare_TLG_fst` directory path.
```

File: wfst_class_lm.md [hunk 8]
```diff
@@ -0,0 +1,159 @@
...
+
+* Generate the lexicon, list of Tokens/units and list of entities from the
```
Comment (LynseyFabel) [inline wfst_class_lm.md:94]:
```suggestion
* Generate the lexicon, the list of tokens/units, and the list of entities from the 
```

File: wfst_class_lm.md [hunk 9]
```diff
@@ -0,0 +1,159 @@
...
+paris
+seattle
+```
+
+## Process to create the `TLG.fst` supporting word classes
+Checkout the `class_lm` branch from https://github.com/anand-nv/riva-asrlib-decoder/tree/class_lm. 
+All the below scripts are available in `src/riva/decoder/scripts/prepare_TLG_fst` path.
+
+* Generate the lexicon, list of Tokens/units and list of entities from the 
+arpa file. This is accomplished using `lexicon_from_arpa.py` script.
```
Comment (LynseyFabel) [inline wfst_class_lm.md:95]:
```suggestion
arpa file. This is accomplished using the `lexicon_from_arpa.py` script. 
```

File: wfst_class_lm.md [hunk 10]
```diff
@@ -0,0 +1,159 @@
...
+Checkout the `class_lm` branch from https://github.com/anand-nv/riva-asrlib-decoder/tree/class_lm. 
+All the below scripts are available in `src/riva/decoder/scripts/prepare_TLG_fst` path.
+
+* Generate the lexicon, list of Tokens/units and list of entities from the 
+arpa file. This is accomplished using `lexicon_from_arpa.py` script. 
+The syntax for the command is as follows:
+
+`lexicon_from_arpa.py --asr_model <path_to_nemo_model> --lm_path <path_to_arpa_file> --grammars <path_to_folder_containing_grammars> <units_txt_path:output> <lexicon_path:output> <classlabels_path:output>`
+
+This generates the list of units(tokens), the lexicon and the list of word
```
Comment (LynseyFabel) [inline wfst_class_lm.md:100]:
```suggestion
This generates the list of units(tokens), the lexicon, and the list of word 
```

File: wfst_class_lm.md [hunk 11]
```diff
@@ -0,0 +1,159 @@
...
+All the below scripts are available in `src/riva/decoder/scripts/prepare_TLG_fst` path.
+
+* Generate the lexicon, list of Tokens/units and list of entities from the 
+arpa file. This is accomplished using `lexicon_from_arpa.py` script. 
+The syntax for the command is as follows:
+
+`lexicon_from_arpa.py --asr_model <path_to_nemo_model> --lm_path <path_to_arpa_file> --grammars <path_to_folder_containing_grammars> <units_txt_path:output> <lexicon_path:output> <classlabels_path:output>`
+
+This generates the list of units(tokens), the lexicon and the list of word 
+classes (to be used as disambiguation symbols)
```
Comment (LynseyFabel) [inline wfst_class_lm.md:101]:
```suggestion
classes (to be used as disambiguation symbols). 
```

File: wfst_class_lm.md [hunk 12]
```diff
@@ -0,0 +1,159 @@
...
+arpa file. This is accomplished using `lexicon_from_arpa.py` script. 
+The syntax for the command is as follows:
+
+`lexicon_from_arpa.py --asr_model <path_to_nemo_model> --lm_path <path_to_arpa_file> --grammars <path_to_folder_containing_grammars> <units_txt_path:output> <lexicon_path:output> <classlabels_path:output>`
+
+This generates the list of units(tokens), the lexicon and the list of word 
+classes (to be used as disambiguation symbols) 
+
+* Generate the FSTs, `T.fst`, `L.fst` and `G.fst`, extend and compose to 
+form `TLG.fst`
```
Comment (LynseyFabel) [inline wfst_class_lm.md:104]:
```suggestion
form `TLG.fst`.
```

File: wfst_class_lm.md [hunk 13]
```diff
@@ -0,0 +1,159 @@
...
+
+`mkgraph_ctc_classlm.sh --units <units_txt_path> --extra_word_disamb <classlabels_path> <lexicon_path>  <path_to_arpa_file> <lang_folder_path:output>`
+
+When run successfully, the above command will generate a graph folder in
+`<lang_folder_path>` with the name `graph_CTC_<lm_name>` which contains
```
Comment (LynseyFabel) [inline wfst_class_lm.md:109]:
```suggestion
`<lang_folder_path>` with the name `graph_CTC_<lm_name>`, which contains
```

File: wfst_class_lm.md [hunk 14]
```diff
@@ -0,0 +1,159 @@
...
+all the files relevant for decoding with WFST
```
Comment (LynseyFabel) [inline wfst_class_lm.md:110]:
```suggestion
all the files relevant for decoding with WFST.
```

File: wfst_class_lm.md [hunk 15]
```diff
@@ -0,0 +1,159 @@
...
+
+
+## Evaluation
+The `evaluate_nemo_wfst.py` script can help evaluate the WFST decoder before 
+being used in the Riva service maker build. This would be useful to
```
Comment (LynseyFabel) [inline wfst_class_lm.md:115]:
```suggestion
being used in the Riva ServiceMaker build. This would be useful to 
```

File: wfst_class_lm.md [hunk 16]
```diff
@@ -0,0 +1,159 @@
...
+`<lang_folder_path>` with the name `graph_CTC_<lm_name>` which contains
+all the files relevant for decoding with WFST
+
+
+## Evaluation
+The `evaluate_nemo_wfst.py` script can help evaluate the WFST decoder before 
+being used in the Riva service maker build. This would be useful to 
+debug any issues with `TLG.fst`
+
+The script can be run as follows
```
Comment (LynseyFabel) [inline wfst_class_lm.md:118]:
```suggestion
The script can be run as follows:
```

File: wfst_class_lm.md [hunk 17]
```diff
@@ -0,0 +1,159 @@
...
+`./evaluate_nemo_wfst.py <model_path> <graph_path> <path_containing_audio_files> <results_txt_file:output>`
+
+## Deploying in Riva
+The generated `TLG.fst` can be used with the kaldi decoder in riva. To build the riva ASR service using the generated
```
Comment (LynseyFabel) [inline wfst_class_lm.md:122]:
```suggestion
The generated `TLG.fst` can be used with the Kaldi decoder in Riva. To build the Riva ASR service using the generated
```

File: wfst_class_lm.md [hunk 18]
```diff
@@ -0,0 +1,159 @@
...
+`TLG.fst` the following command can be used
```
Comment (LynseyFabel) [inline wfst_class_lm.md:123]:
```suggestion
`TLG.fst`, run:
```

**Review by LynseyFabel — APPROVED** [review]
I've left quite a few editorial comments. I've approved assuming they will be implemented. Thank you.

**Review by LynseyFabel — APPROVED** [review]
I've left quite a few editorial comments. I've approved assuming they will be implemented. Thank you.

**Comment by anand-nv** [comment]
> please review @LynseyFabel, this is needed in the release. @anand-nv there should be a link to these tutorials in the Riva docs on the left pane under `Tutorials -> Speech Recognition` section, right?

Yes. Am creating one.

**Comment by anand-nv** [comment]
> please review @LynseyFabel, this is needed in the release. @anand-nv there should be a link to these tutorials in the Riva docs on the left pane under `Tutorials -> Speech Recognition` section, right?

Yes. Added the links

**Comment by anand-nv** [comment]
> I've left quite a few editorial comments. I've approved assuming they will be implemented. Thank you.

Thanks. They've been implemented.

**Pull request merged**